File: src/main/java/dev/iancmd/operationrelife/handlers/ModEventHandler.java
package dev.iancmd.operationrelife.handlers;

import dev.iancmd.operationrelife.init.ModBlocks;
import dev.iancmd.operationrelife.init.ModItems;
import net.minecraft.client.renderer.ItemBlockRenderTypes;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.event.CreativeModeTabEvent;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.apache.logging.log4j.Level;

import static dev.iancmd.operationrelife.core.OperationRelife.MODID;
import static dev.iancmd.operationrelife.core.OperationRelife.getLogger;

public class ModEventHandler {
  private static final IEventBus BUS = FMLJavaModLoadingContext.get().getModEventBus();
  public static void registerAll() {
    ModEventHandler.BUS.register(ModEventHandler.class);
    ModEventHandler.BUS.register(new ModEventHandler());
  }
  @SubscribeEvent
  public static void onCommonSetup(FMLCommonSetupEvent event) {
    ForgeEventHandler.registerEvents();
  }
  @SubscribeEvent
  public static void onClientSetup(FMLClientSetupEvent event) {
    ItemBlockRenderTypes.setRenderLayer(ModBlocks.BlockRegistry.BLUE_SOUL_HOLDER.get(), RenderType.translucent());
  }
  @SubscribeEvent
  public void onRegisterCreativeModeTabs(CreativeModeTabEvent.Register event) {
    getLogger().log(Level.DEBUG, "Registering creative mode tabs...");
    event.registerCreativeModeTab(new ResourceLocation(MODID, "creative_mod_tab"), (builder) -> builder
              .title(Component.translatable("itemGroup." + MODID + ".creative_mod_tab")).
              icon(() -> ModItems.ItemRegistry.RED_SOUL_HOLDER.get().getDefaultInstance())
              .displayItems((p_259897_, output, p_259072_) -> {
                output.accept(ModItems.ItemRegistry.BLUE_SOUL_HOLDER.get());
                output.accept(ModItems.ItemRegistry.RED_SOUL_HOLDER.get());
              }));
  }
}

File: src/main/kotlin/dev/iancmd/operationrelife/handler/ModEventHandler.kt
package dev.iancmd.operationrelife.handler

import dev.iancmd.operationrelife.core.OperationRelife.Companion.getLogger
import dev.iancmd.operationrelife.init.ModItems
import net.minecraft.world.item.CreativeModeTabs
import net.minecraftforge.event.BuildCreativeModeTabContentsEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent
import org.apache.logging.log4j.Level.DEBUG
import thedarkcolour.kotlinforforge.forge.MOD_BUS

class ModEventHandler {
  @SubscribeEvent
  fun onRegisterCreativeModeTabs(event: BuildCreativeModeTabContentsEvent) {
    getLogger().log(DEBUG, "Registering creative mode tabs...")
    if (event.tabKey === CreativeModeTabs.FUNCTIONAL_BLOCKS) {
      event.accept(ModItems.ItemRegistry.BLUE_SOUL_HOLDER)
      event.accept(ModItems.ItemRegistry.RED_SOUL_HOLDER)
    }
  }
  @SubscribeEvent
  @Suppress("UNUSED_PARAMETER")
  fun onClientSetup(event: FMLClientSetupEvent) {
    ForgeEventHandler.registerAll()
  }
  companion object {
    fun registerAll() {
      getLogger().log(DEBUG, "Registering mod event handler...")
      MOD_BUS.register(ModEventHandler())
    }
  }
}
